### DISCUSSION TITLE
Some usability suggestions

### DISCUSSION THREAD

**Discussion by Fornax96** [Syncplay/syncplay#448] [Archive]
Hello, I have used Syncplay a few times to watch movies with my friends during quarantine. I think it's a great piece of software! But I have noticed that there are a few things about the UI that always confuses people. In this issue I will put some potential fixes for these issues.

This is what people see when first starting Syncplay:
![image](https://user-images.githubusercontent.com/1247928/107350306-94a8e700-6ac9-11eb-8cbb-c440ba23f1e0.png)

 * *Server address*: Would it be possible to give the servers friendly names? Port numbers are hard to recognize. It would be easy if I could just tell my friends "We are on server Pluto" or something like that.
 * *Path to media player*: Most people don't know where their media player is installed. The Browse button stands out more than the drop down menu, so people are not noticing that it's there. It would be better to show a radio button list of installed media players with their icons. And add a button on the side to select a custom media player path. Also, since mpv is free and open source, would it be possible to bundle it with this software so people don't have to install it separately? 
 * *Path to video*: It's easier to select this from the playlist when you're already in the room, doesn't have to be in the settings menu if it were up to me.
 * *Player arguments*: Most people don't know what this means and don't need it either. This could be moved to the Misc menu to remove clutter.

Then there's a button "Run Syncplay". Isn't Syncplay already running? I'd put something like "Join room" on this button instead.

So you click "Run Syncplay" and you see this:
![image](https://user-images.githubusercontent.com/1247928/107351827-60cec100-6acb-11eb-904b-95d84240986a.png)

 * *Selecting a video to play*: It's a lot of work to make sure everyone has the same media file. To work around this I usually upload the video I want to watch to my own webserver. Then I use the File > Open media stream URL option to add it to the player. It's never clear that you can click the URL next to my username to play the same video. It would be handy if the same video is always selected for everyone in the room. Just use the first item in the playlist for that.
 * *Enable shared playlists*: Isn't the point of this software to watch videos together? Why would you want to turn this off? I'd move this option to a drop-down menu to reduce clutter.
 * *Open media file / stream URL*: These options are tucked away in the File menu, while these are arguably the most important functions of the software. I would add a text field above the playlist (where the shared playlists button is now) where you can paste an URL / path and press Enter to add it to the playlist. This would make it instantly clear to people how you can add videos to the playlist.

I hope this isn't too overwhelming. Do whatever you want with my feedback :)

**Comment by daniel-123** [comment]
Your screenshot doesn't show what users see when first running Syncplay - the "Show more settings" is off by default, which presents much more streamlined interface. Notably this hides both player arguments and path to video.

Server names are very plain hostname and optional port and have dropdown pre-populated with servers we host ourselves (until recently there was no pre-populated list at all). When it comes to aliases I'd say that they would be much more confusing than what's there now. Especially given that Syncplay has been using this way of referring to servers for almost a decade by now.

Path to media player should in most circumstances automatically populate itself with sane default. So there should be no need to touch it in first place. I'm obviously biased by familiarity with existing interface, but I don't really see the dropdown/browse button as confusing. It's also pretty standard for file dialogs across various operating systems.

When it comes to the "main" interface I generally have quite different workflow. For example I basically *never* use File -> Open, but instead either drag and drop files/links to the playlist area or start Syncplay by selecting "Open with..." from context menu of a video file. I switch files either by clicking on them on the playlist or by dragging them into the player window itself. That said I could see valid reasons for adding some buttons next to the playlist window.

Ability to enable/disable shared playlists is there mostly for historical reasons - for pretty long time it was a somewhat wonky feature. Even now it doesn't work perfectly. Originally it made a lot of sense to have ability to turn it on/off at a glance. If we were to add some controls to the playlist interface I would probably still be in favor of leaving it there.

As a side note - most of the UX decisions were made after decent amount of deliberation so there almost always is *some* decent reason why every bit of it is the way it is. That said the interface likely has accumulated a bit of cruft over the years. On another hand though - what currently is there tends to work well enough, so nobody from core developers is really bothered to spend a lot of effort on changing it.

**Comment by Et0h** [comment]
Hi, thanks for your suggestions. I don't see any of your proposed changes as likely to be implemented for the reasons Dan said. However, it is possible that a few improvements inspired by your suggestions could be helpful. These are:

- Adding the playlist options as a section to the menu bar
- Adding buttons for "Set Media Directories", "Add File(s)", "Add URL(s)" beneath the playlist

The former improvement I think has no downsides. The latter adds clutter, but does make the most obvious options for Syncplay available. The new buttons could be something which people could choose to disable if they wished, in the same way that the playback and auto-play buttons can be toggled.

When you talk about people being confused, are you talking about your own experience of people being confused or seeing others be confused, or are you talking about a theoretical person being confused? Also, did the people who are confused read the guide at https://syncplay.pl/guide/client/ and how quickly did they figure things out?

I'll comment on some of your suggestions.

> This is what people see when first starting Syncplay:

As Dan says, this is not what people see when they first start Syncplay.

> * _Server address_: Would it be possible to give the servers friendly names? Port numbers are hard to recognize. It would be easy if I could just tell my friends "We are on server Pluto" or something like that.

If we change system then this would be confusing with anyone still using old versions, or people who already advertised their server based on a port. It is also unclear how it would work with the CLI, and it might be harder to copy and paste. By using IP and port we are following a convention based on standards, and anything else could be confusing.

> * _Path to media player_: Most people don't know where their media player is installed.
> The Browse button stands out more than the drop down menu, so people are not noticing that it's there. It would be better to show a radio button list of installed media players with their icons. And add a button on the side to select a custom media player path.

Syncplay is designed to automatically locate media players in their default directories. If you know of any paths we missed then let us know. Choosing the player via the path is in line with what we do for CLI mode, and if we had to diverge then that could make things overly complicated. This set-up also allows people to have multiple versions of a media players without issue (e.g. the stable and development versions of mpv).


>Also, since mpv is free and open source, would it be possible to bundle it with this software so people don't have to install it separately?

If we install is, then we have to update it - so that then becomes an additional maintenance burden for us. Also, just because two pieces of software are open source does not mean their licenses are compatible (although I think mpv's is compatible with Syncplay last I checked).


> * _Path to video_: It's easier to select this from the playlist when you're already in the room, doesn't have to be in the settings menu if it were up to me.

That's why it's not visible by default.

> * _Player arguments_: Most people don't know what this means and don't need it either. This could be moved to the Misc menu to remove clutter.

Again, that is why this is not visible by default (unless you choose to open a file with Syncplay). For those who do use it, we want it to be prominent so people know what options are currently set. Putting it under misc is confusing because these are per-player arguments.

> 
> Then there's a button "Run Syncplay". Isn't Syncplay already running? I'd put something like "Join room" on this button instead.

It is arguably the Syncplay configuration window which is running, not the main Syncplay client. We could change the text, but we already have the text we have and existing users are already used to it. Being in the bottom right hand corner should make it obvious that this is the "Let's go" button.

> * _Selecting a video to play_: It's a lot of work to make sure everyone has the same media file. To work around this I usually upload the video I want to watch to my own webserver. Then I use the File > Open media stream URL option to add it to the player. It's never clear that you can click the URL next to my username to play the same video. It would be handy if the same video is always selected for everyone in the room. Just use the first item in the playlist for that.

I never use File -> Open Media (or the media open option of my media player for that matter) and very rarely double click on names, I nearly always just add media to the playlist which will auto-play it. I occasionally drag a file into my player to load it. The double click feature was implemented before the shared playlist feature existed, and is now mostly redundant.

> * _Enable shared playlists_: Isn't the point of this software to watch videos together? Why would you want to turn this off? I'd move this option to a drop-down menu to reduce clutter.

If it turns out you have a bad copy of a video then you can disable the shared playlist to manually switch to an alternative version of the video without it automatically switching back. You can then quickly turn it back on. It doesn't actually take up any space, and people are already be used to it in its current location.

> * _Open media file / stream URL_: These options are tucked away in the File menu, while these are arguably the most important functions of the software. I would add a text field above the playlist (where the shared playlists button is now) where you can paste an URL / path and press Enter to add it to the playlist. This would make it instantly clear to people how you can add videos to the playlist.

I hardly ever open media, I add files/URLs to the playlist instead. People who want to add stuff can do so via the /qa command in the chat box as per https://syncplay.pl/guide/client/
